### PR TITLE
feat: reward initiator for ECDSAServiceManagerBase

### PR DESCRIPTION
**Pull request by Gajesh2007**

### CODE REVIEW

File: src/unaudited/ECDSAServiceManagerBase.sol [hunk 1]
```diff
@@ -234,6 +249,20 @@ abstract contract ECDSAServiceManagerBase is
...
+    }
+
+    function _setRewardsInitiator(address newRewardsInitiator) internal {
+        emit RewardsInitiatorUpdated(rewardsInitiator, newRewardsInitiator);
+        rewardsInitiator = newRewardsInitiator;
+    }
+
     // storage gap for upgradeability
     // slither-disable-next-line shadowing-state
     uint256[50] private __GAP;
```
Comment (stevennevins) [inline src/unaudited/ECDSAServiceManagerBase.sol]: Need to update the storage gap for the new state

File: src/unaudited/ECDSAServiceManagerBase.sol [hunk 2]
```diff
@@ -42,6 +41,17 @@ abstract contract ECDSAServiceManagerBase is
         _;
     }
 
+    /**
+     * @dev Ensures that the function is only callable by the `rewardsInitiator`.
+     */
+    modifier onlyRewardsInitiator() {
+        require(
```
Comment (stevennevins) [inline src/unaudited/ECDSAServiceManagerBase.sol:53]: Can you put this in an internal function

```
modifier onlyRewardInitiator() {
    _checkRewardsInitiator();
    _;
}
```

File: src/unaudited/ECDSAServiceManagerBase.sol [hunk 3]
```diff
@@ -168,9 +181,11 @@ abstract contract ECDSAServiceManagerBase is
                 address(this),
                 rewardsSubmissions[i].amount
             );
+            uint256 allowance =
+                rewardsSubmissions[i].token.allowance(address(this), rewardsCoordinator);
             rewardsSubmissions[i].token.approve(
                 rewardsCoordinator,
-                rewardsSubmissions[i].amount
+                rewardsSubmissions[i].amount + allowance
```
Comment (stevennevins) [inline src/unaudited/ECDSAServiceManagerBase.sol:192]: nice thanks for doing this as well

File: src/unaudited/ECDSAServiceManagerBase.sol (hunk 1 above)
Comment (Gajesh2007) [reply src/unaudited/ECDSAServiceManagerBase.sol]: by how much?

File: src/unaudited/ECDSAServiceManagerBase.sol (hunk 2 above)
Comment (Gajesh2007) [reply src/unaudited/ECDSAServiceManagerBase.sol:53]: it's just one require() check. do u think we need to put it in another func?

File: src/unaudited/ECDSAServiceManagerBase.sol (hunk 1 above)
Comment (stevennevins) [reply src/unaudited/ECDSAServiceManagerBase.sol]: Needs to be 49 now.  

Basically you allocate up front 50 slots so when you make storage changes you don't shift the storage of contracts in the inheritance graph. Then for each storage slot you use in the contract you deduct 1 from your gap

File: src/unaudited/ECDSAServiceManagerBase.sol (hunk 2 above)
Comment (stevennevins) [reply src/unaudited/ECDSAServiceManagerBase.sol:53]: Yeah that way if we go to add the modifier in the future we don't forgot to make this change to reduce the duplication of the inlined code

**Comment by Gajesh2007** [comment]
ready for merge @stevennevins

**Comment by stevennevins** [comment]
lgtm ty

**Review by stevennevins — APPROVED** [review]

**Pull request merged**